src/ReverseProxy/Health/ActiveHealthCheckMonitor.cs

```
@@ -114,87 +112,67 @@ private async Task ProbeCluster(ClusterState cluster)
-                    }
-                    catch (Exception ex)
-                    {
-                        // Suppress exceptions to ensure all responses get a chance to be disposed.
-                        Log.ErrorOccuredDuringActiveHealthProbingShutdownOnCluster(_logger, cluster.ClusterId, ex);
-                    }
-                    finally
-                    {
-                        // Dispose CancellationTokenSource even if the response task threw an exception.
-                        // Dispose() is not expected to throw here.
-                        probeTask.Cts.Dispose();
-                    }
+        private async Task<DestinationProbingResult> ProbeDestinationAsync(ClusterState cluster, DestinationState destination, TimeSpan timeout)
+        {
+            var cts = new CancellationTokenSource(timeout);
+            var waitingForSendAsync = false;
+            try
+            {
+                var request = _probingRequestFactory.CreateRequest(cluster.Model, destination.Model);
+                Log.SendingHealthProbeToEndpointOfDestination(_logger, request.RequestUri, destination.DestinationId, cluster.ClusterId);
+
+                var responseTask = cluster.Model.HttpClient.SendAsync(request, cts.Token);
+                waitingForSendAsync = true;
+
+                var response = await responseTask;
+                Log.DestinationProbingCompleted(_logger, destination.DestinationId, cluster.ClusterId, (int)response.StatusCode);
+
+                return new DestinationProbingResult(destination, response, null);
+            }
+            catch (Exception ex)
+            {
+                if (waitingForSendAsync)
+                {
+                    Log.DestinationProbingFailed(_logger, destination.DestinationId, cluster.ClusterId, ex);
+                }
+                else
+                {
+                    Log.ActiveHealthProbeConstructionFailedOnCluster(_logger, destination.DestinationId, cluster.ClusterId, ex);
```

Previously if `SendAsync` failed synchronously, the `DestinationProbingResult` for that destination would not be reported to `ProbingCompleted`.
This was done intentionally, but I don't understand why. @tratcher do you remember?

I think it should be the responsibility of `IActiveHealthCheckPolicy` to decide what to do with a failed request, and the monitor shouldn't be hiding failures.

---

Secondly, is there a benefit to reporting different logs for sync vs async probe failure?

---

I think ActiveHealthProbeConstructionFailedOnCluster is trying to report a failure from _probingRequestFactory.CreateRequest, not cluster.Model.HttpClient.SendAsync. That would make sense to report separately. Sync and async HttpClient exceptions should be reported the same.

---

> If _probingRequestFactory.CreateRequest throws should that probe be reported to IActiveHealthCheckPolicy?

Yes, I think so. It indicates something is seriously wrong with health check logic if we couldn't even create the request.